tree/treeplayer/src/TTreeReader.cxx

```diff
@@ -354,108 +454,76 @@ TTreeReader::EEntryStatus TTreeReader::SetEntryBase(Long64_t entry, Bool_t local
-                 "The current tree in the TChain %s has changed (e.g. by TTree::Process) "
-                 "even though TTreeReader::SetEntry() was called, which switched the tree "
-                 "again. Did you mean to call TTreeReader::SetLocalEntry()?",
-                 fTree->GetName());
+               "There was an issue opening the last file associated to the TChain "
+               "being processed.");
+         fEntryStatus = kEntryChainFileError;
+         return fEntryStatus;
       }
-   }
 
-   if (fDirector->GetTree() != fTree->GetTree()
-       || fMostRecentTreeNumber != fTree->GetTreeNumber()) {
-      fDirector->SetTree(fTree->GetTree());
-      if (fProxiesSet) {
-         for (auto value: fValues) {
-            value->NotifyNewTree(fTree->GetTree());
+      if (loadResult == -2) {
+         fDirector->SetTree(fTree->GetTree());
+         if (fProxiesSet) {
+            for (auto value: fValues) {
+               value->NotifyNewTree(fTree->GetTree());
+            }
          }
+         fEntryStatus = kEntryNotFound;
+         return fEntryStatus;
       }
-   }
-
-   fMostRecentTreeNumber = fTree->GetTreeNumber();
-
-   if (!fProxiesSet) {
-      // Tell readers we now have a tree.
-      // fValues gets insertions during this loop (when parameterized arrays are read),
-      // invalidating iterators. Use old-school counting instead.
-      for (size_t i = 0; i < fValues.size(); ++i) {
-         ROOT::Internal::TTreeReaderValueBase* reader = fValues[i];
-         reader->CreateProxy();
-         if (!reader->GetProxy()){
-            fEntryStatus = kEntryDictionaryError;
-            return fEntryStatus;
-         }
 
+      if (loadResult == -1) {
+         // The chain is empty
+         fEntryStatus = kEntryNotFound;
+         return fEntryStatus;
       }
-      // If at least one proxy was there and no error occurred, we assume the proxies to be set.
-      fProxiesSet = !fValues.empty();
-
-      // Now we need to properly set the TTreeCache. We do this in steps:
-      // 1. We set the entry range according to the entry range of the TTreeReader
-      // 2. We add to the cache the branches identifying them by the name the user provided
-      //    upon creation of the TTreeReader{Value, Array}s
-      // 3. We stop the learning phase.
-      // Operations 1, 2 and 3 need to happen in this order. See: https://sft.its.cern.ch/jira/browse/ROOT-9773?focusedCommentId=87837
-      if (fProxiesSet) {
-         const auto curFile = fTree->GetCurrentFile();
-         if (curFile && fTree->GetTree()->GetReadCache(curFile, true)) {
-            if (!(-1LL == fEndEntry && 0ULL == fBeginEntry)) {
-               // We need to avoid to pass -1 as end entry to the SetCacheEntryRange method
-               const auto lastEntry = (-1LL == fEndEntry) ? fTree->GetEntriesFast() : fEndEntry;
-               fTree->SetCacheEntryRange(fBeginEntry, lastEntry);
-            }
+
+      if (loadResult == -4) {
```

Comment:
Shouldn't this be an `enum` for the sake of clarity?

Reply:
correct but out of the scope of this PR.

Reply:
correct but out of the scope of this PR.